ocaml/typing/typecore.ml

```diff
@@ -6505,6 +6512,18 @@ and type_application env app_loc expected_mode position funct funct_mode sargs r
       in
       submode ~loc:app_loc ~env ~reason:(Application ty_ret)
         mode_ret expected_mode;
+      (match maybe_outer_ret_mode with
```

Comment:
This is currently ignoring the `let position = if partial_app then ...` above. I think it would be better if we only applied this constraint for genuine tail calls, which requires taking that into account.